chore: multipart suggestions for let_unit_value lint

scottgerring opened the pull request:
This should address #13099 for the let_unit test. 

changelog: [let_unit]: Updated let_unit to use multipart_suggestions where appropriate

scottgerring commented:
Flip1995 has the context of these, so if they have time: 

r? flip1995 

(But this is still a draft!)

scottgerring commented on tests/ui/let_unit.stderr:
```diff
@@ -52,23 +52,17 @@ LL +     };
-help: variable `res` of type `()` can be replaced with explicit `()`
+LL ~         returns_unit();
+LL ~         returns_result(()).unwrap();
+LL ~         returns_result(()).unwrap();
```
It jumps out at me that here:

https://github.com/rust-lang/rust-clippy/blob/fced7ed109d97661a399ee5577af16d51079e12f/tests/ui/let_unit.rs#L187-L195

we're now suggesting changing both L190 and L191, which I believe is correct.

flip1995 commented on clippy_lints/src/unit_types/let_unit_value.rs:
```diff
@@ -71,25 +71,46 @@ pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, local: &'tcx LetStmt<'_>) {
-                        });
+                    // If this is a binding pattern, we need to add suggestions to remove any usages
+                    // of the variable
+                    if let PatKind::Binding(_, binding_hir_id, ident, ..) = local.pat.kind {
+                        if let Some(body_id) = cx.enclosing_body.as_ref() {
```
This should still be a let-chain:

```suggestion
                    if let PatKind::Binding(_, binding_hir_id, ident, ..) = local.pat.kind
                        && let Some(body_id) = cx.enclosing_body.as_ref() {
```

This will change the indentation of all the code below.

flip1995 commented on clippy_lints/src/unit_types/let_unit_value.rs:
```diff
@@ -71,25 +71,46 @@ pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, local: &'tcx LetStmt<'_>) {
+                            walk_body(&mut visitor, body);
+
+                            // Add suggestions for replacing variable usages
+                            let has_usages = !visitor.spans.is_empty();
+                            if has_usages {
+                                suggestions.extend(visitor.spans.into_iter().map(|span| (span, "()".to_string())));
+                            }
```
Why the extra check for `is_empty`?

```suggestion
                            suggestions.extend(visitor.spans.into_iter().map(|span| (span, "()".to_string())));
```

`extend` with an empty iterator will just do nothing.

flip1995 reviewed (approved):
Great work! Just 2 codestyle comments. 

Please also rebase your PR on top of the current master. We changed our CI recently and GitHub isn't good at picking this up on older PRs.

scottgerring commented:
Hey @flip1995 thanks for the review! I rebased and addressed the resulting dogfooding lints and your own comments, so there's a bit of a delta on top of what you reviewed. Should be good to go, though, I reckon.

scottgerring commented on clippy_lints/src/unit_types/let_unit_value.rs:
```diff
@@ -71,26 +72,48 @@ pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, local: &'tcx LetStmt<'_>) {
                     }
+
+                    // Emit appropriate diagnostic based on whether there are usages of the let binding
+                    match suggestions.len().cmp(&1) {
```
Yep, that's way less awful. Let me clean it up 👍

scottgerring commented:
@flip1995 that should be it! 🙏

scottgerring commented:
Hey @flip1995 , do you have a chance to have a quick look at this? We're very close to finishing the whole issue!

flip1995 commented:
Sorry for not replying. I won't get to review any Clippy PRs until around christmas... I'll reassign

r? clippy

blyxyas commented on clippy_lints/src/unit_types/let_unit_value.rs:
```diff
@@ -71,25 +71,38 @@ pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, local: &'tcx LetStmt<'_>) {
+                    if let PatKind::Binding(_, binding_hir_id, ..) = local.pat.kind
                         && let Some(body_id) = cx.enclosing_body.as_ref()
-                        && let body = cx.tcx.hir().body(*body_id)
-                        && is_local_used(cx, body, binding_hir_id)
```
This line is completely removed, is this intended?

blyxyas replied:
And sorry for the late review, I've opened this PR maybe 5 times and never got around to clicking that "Submit review" button

scottgerring replied:
Hey @blyxyas no worries!

I think it is unnecessary now; this part here collects usages:

https://github.com/rust-lang/rust-clippy/blob/ec243887d3524545922e340a4e55c4bd174e966e/clippy_lints/src/unit_types/let_unit_value.rs#L90-L92

... and we then emit a lint based on whether or not we had any: 

https://github.com/rust-lang/rust-clippy/blob/ec243887d3524545922e340a4e55c4bd174e966e/clippy_lints/src/unit_types/let_unit_value.rs#L99-L105

It could be added back as a sort of "performance optimization guard" to remove the need to call `walk_body`, assuming that `is_local_used` is quicker, but it doesn't change the behaviour (I also double checked this!)

blyxyas replied:
I've been all mornig trying to measure this but it's in a very awkward spot to get a hold of how many heavy-lifting that function does.

Being that in the case that the body is visited is already in the linting-case, let's just ignore it.

blyxyas reviewed (approved):
LGTM, thanks! :heart:
